server/src/main/java/org/elasticsearch/index/mapper/IgnoredValuesFieldMapper.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.util.ByteUtils;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * Mapper for the {@code _ignored_values} field.
+ *
+ * A field mapper that records fields that have been ignored, along with their values. It's intended for use
+ * in indexes with synthetic source to reconstruct the latter, taking into account fields that got ignored during
+ * indexing.
+ *
+ * This overlaps with {@link IgnoredFieldMapper} that tracks just the ignored field names. It's worth evaluating
+ * if we can replace it for all use cases to avoid duplication, assuming that the storage tradeoff is favorable.
+ */
+public class IgnoredValuesFieldMapper extends MetadataFieldMapper {
+
+    private static final int OFFSET_IN_NAME = 100_000;
+
+    public static final String NAME = "_ignored_values";
+
+    public static final IgnoredValuesFieldMapper INSTANCE = new IgnoredValuesFieldMapper();
+
+    public static final TypeParser PARSER = new FixedTypeParser(context -> INSTANCE);
+
+    static final NodeFeature TRACK_IGNORED_VALUES = new NodeFeature("mapper.track_ignored_values");
+
+    /*
+     * Container for the ignored field data:
+     *  - the full name
+     *  - the offset in the full name indicating the end of the substring matching
+     *    the full name of the parent field
+     *  - the value, encoded as a byte array
+     */
+    public record Values(String name, int parentOffset, BytesRef value) {}
+
+    static final class IgnoredValuesFieldMapperType extends StringFieldType {
+
+        private static final IgnoredValuesFieldMapperType INSTANCE = new IgnoredValuesFieldMapperType();
+
+        private IgnoredValuesFieldMapperType() {
+            super(NAME, false, true, false, TextSearchInfo.NONE, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return NAME;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
+            return new StoredValueFetcher(context.lookup(), NAME);
+        }
+    }
+
+    private IgnoredValuesFieldMapper() {
+        super(IgnoredValuesFieldMapperType.INSTANCE);
+    }
+
+    @Override
+    protected String contentType() {
+        return NAME;
+    }
+
+    @Override
+    public void postParse(DocumentParserContext context) {
+        for (Values values : context.getIgnoredFieldValues()) {
+            context.doc().add(new StringField(NAME, encode(values), Field.Store.YES));
```

Comment:
Should we use `StoredField`? We don't need this to be indexed, right?

Reply:
The type constructor sets it as non-indexed, non-docvalues so this seems equivalent?

Reply:
`StoredField` will also allow you to use plain bytes.

Reply:
I was referring to this in my initial comment https://github.com/apache/lucene/blob/d57267a0e6061eaee549b9b56f9d1c77ece83a1a/lucene/core/src/java/org/apache/lucene/document/StringField.java#L23.

Reply:
Nice, I can just use it here. Done.